File: ravewave-ios/src/scene/sceneRandomizer.ts
import { AnalyzerFrame } from '../analyzer/AnalyzerEngine';
import { ColorModes, PostEffects, SceneState, VisualLayers } from './sceneState';

const randomFloat = (min: number, max: number): number => min + Math.random() * (max - min);
const randomInt = (min: number, max: number): number => Math.floor(randomFloat(min, max + 1));

const shuffled = <T,>(items: readonly T[]): T[] => {
  const copy = [...items];
  for (let i = copy.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
};

const pickSubset = <T,>(items: readonly T[], min: number, max: number): Set<T> => {
  const count = randomInt(Math.min(min, items.length), Math.min(max, items.length));
  return new Set(shuffled(items).slice(0, count));
};

const clamp = (value: number, min: number, max: number): number => Math.max(min, Math.min(max, value));

export type EvolvePulseState = {
  lastMutationAt: number;
  phraseMomentum: number;
};

export const createEvolvePulseState = (): EvolvePulseState => ({
  lastMutationAt: 0,
  phraseMomentum: 0,
});

export const randomizeScene = (scene: SceneState): SceneState => ({
  ...scene,
  enabledLayers: pickSubset(VisualLayers, 1, VisualLayers.length),
  enabledEffects: pickSubset(PostEffects, 0, Math.max(0, PostEffects.length - 2)),
  colorMode: ColorModes[randomInt(0, ColorModes.length - 1)],
  fxIntensity: Number(randomFloat(0.25, 1).toFixed(2)),
  speed: Number(randomFloat(0.2, 1).toFixed(2)),
  tileCount: randomInt(2, 6),
  symmetrySegments: randomInt(4, 12),
});

const toggleLayer = (layers: Set<(typeof VisualLayers)[number]>, addBias: number) => {
  const pool = shuffled(VisualLayers);
  const shouldAdd = layers.size <= 2 || (layers.size < 12 && Math.random() < addBias);
  if (shouldAdd) {
    const next = pool.find((layer) => !layers.has(layer));
    if (next) layers.add(next);
    return;
  }
  const removable = [...layers].filter((_, index, arr) => arr.length > 1);
  if (removable.length > 0) {
    layers.delete(removable[randomInt(0, removable.length - 1)]);
  }
};

const toggleEffect = (effects: Set<(typeof PostEffects)[number]>, addBias: number) => {
  const pool = shuffled(PostEffects);
  const shouldAdd = effects.size < 7 && Math.random() < addBias;
  if (shouldAdd) {
    const next = pool.find((effect) => !effects.has(effect));
    if (next) effects.add(next);
    return;
  }
  if (effects.size > 0) {
    const enabled = [...effects];
    effects.delete(enabled[randomInt(0, enabled.length - 1)]);
  }
};

export const evolveSceneWithAudio = (
  scene: SceneState,
  frame: AnalyzerFrame,
  pulseState: EvolvePulseState
): SceneState => {
  const now = frame.timestamp;
  const accent = frame.beatPulse * 0.55 + frame.onsetPulse * 0.35 + frame.attack * 0.1;
  pulseState.phraseMomentum = clamp(
    pulseState.phraseMomentum * 0.82 + frame.energy * 0.12 + frame.beatPulse * 0.22 + frame.onsetPulse * 0.18,
    0,
    1.8
  );

  const minInterval = 220 + (1 - accent) * 480;
  if (now - pulseState.lastMutationAt < minInterval) {
    return scene;
  }

  const mutationChance = accent * 0.9 + frame.energy * 0.18;
  if (Math.random() > mutationChance) {
    return scene;
  }

  pulseState.lastMutationAt = now;

  const nextLayers = new Set(scene.enabledLayers);
  const nextEffects = new Set(scene.enabledEffects);
  let nextColorMode = scene.colorMode;
  let nextFxIntensity = scene.fxIntensity;
  let nextSpeed = scene.speed;
  let nextTileCount = scene.tileCount;
  let nextSymmetry = scene.symmetrySegments;

  const phraseHit = pulseState.phraseMomentum > 1.15 && frame.beatPulse > 0.58;
  const mutationCount = phraseHit ? randomInt(2, 4) : accent > 0.72 ? randomInt(1, 3) : 1;

  for (let i = 0; i < mutationCount; i += 1) {
    const roll = Math.random();
    if (roll < 0.52) {
      toggleLayer(nextLayers, 0.74 - frame.decay * 0.2);
    } else if (roll < 0.8) {
      toggleEffect(nextEffects, 0.58 + frame.highs * 0.16);
    } else if (roll < 0.9) {
      if (phraseHit || Math.random() < frame.highs * 0.5) {
        nextColorMode = ColorModes[randomInt(0, ColorModes.length - 1)];
      }
    } else {
      nextFxIntensity = Number(clamp(nextFxIntensity + (frame.attack - frame.decay) * 0.08 + randomFloat(-0.04, 0.04), 0, 1).toFixed(2));
      nextSpeed = Number(clamp(nextSpeed + frame.beatPulse * 0.05 - frame.decay * 0.03 + randomFloat(-0.03, 0.03), 0, 1).toFixed(2));
      nextTileCount = clamp(nextTileCount + (phraseHit ? randomInt(-1, 1) : 0), 2, 6);
      nextSymmetry = clamp(nextSymmetry + (frame.highs > 0.62 && Math.random() < 0.4 ? randomInt(-1, 1) : 0), 4, 12);
    }
  }

  if (phraseHit) {
    pulseState.phraseMomentum *= 0.58;
  }

  return {
    ...scene,
    enabledLayers: nextLayers.size > 0 ? nextLayers : new Set([VisualLayers[randomInt(0, VisualLayers.length - 1)]]),
    enabledEffects: nextEffects,
    colorMode: nextColorMode,
    fxIntensity: nextFxIntensity,
    speed: nextSpeed,
    tileCount: nextTileCount,
    symmetrySegments: nextSymmetry,
  };
};

export const nextEvolveDelayMs = (frame: AnalyzerFrame): number => {
  const accent = frame.beatPulse * 0.6 + frame.onsetPulse * 0.25 + frame.energy * 0.15;
  return Math.round(180 + (1 - clamp(accent, 0, 1)) * 320);
};

File: app/src/main/java/com/ravewave/app/scene/SceneRandomizer.kt
package com.ravewave.app.scene

import com.ravewave.app.analyzer.AnalyzerMetrics
import kotlin.math.roundToInt
import kotlin.random.Random

object SceneRandomizer {
    data class EvolvePulseState(
        var lastMutationAt: Long = 0L,
        var phraseMomentum: Float = 0f
    )

    private fun randomFloat(min: Float, max: Float): Float = min + Random.nextFloat() * (max - min)
    private fun randomInt(min: Int, max: Int): Int = Random.nextInt(from = min, until = max + 1)
    private fun clamp(value: Float, min: Float, max: Float): Float = value.coerceIn(min, max)

    private fun <T> pickSubset(items: List<T>, min: Int, max: Int): Set<T> {
        val count = randomInt(min.coerceAtMost(items.size), max.coerceAtMost(items.size))
        return items.shuffled().take(count).toSet()
    }

    fun randomizeScene(scene: SceneState): SceneState {
        return scene.copy(
            enabledLayers = pickSubset(VisualLayer.entries, 1, VisualLayer.entries.size),
            enabledEffects = pickSubset(PostEffect.entries, 0, (PostEffect.entries.size - 2).coerceAtLeast(0)),
            colorMode = ColorMode.entries[randomInt(0, ColorMode.entries.size - 1)],
            fxIntensity = (randomFloat(0.25f, 1f) * 100f).roundToInt() / 100f,
            speed = (randomFloat(0.2f, 1f) * 100f).roundToInt() / 100f,
            tileCount = randomInt(2, 6),
            symmetrySegments = randomInt(4, 12)
        )
    }

    private fun toggleLayer(layers: MutableSet<VisualLayer>, addBias: Float) {
        val shouldAdd = layers.size <= 2 || (layers.size < 12 && Random.nextFloat() < addBias)
        if (shouldAdd) {
            VisualLayer.entries.shuffled().firstOrNull { !layers.contains(it) }?.let(layers::add)
            return
        }
        if (layers.size > 1) {
            layers.remove(layers.shuffled().first())
        }
    }

    private fun toggleEffect(effects: MutableSet<PostEffect>, addBias: Float) {
        val shouldAdd = effects.size < 7 && Random.nextFloat() < addBias
        if (shouldAdd) {
            PostEffect.entries.shuffled().firstOrNull { !effects.contains(it) }?.let(effects::add)
            return
        }
        if (effects.isNotEmpty()) {
            effects.remove(effects.shuffled().first())
        }
    }

    fun evolveSceneWithAudio(
        scene: SceneState,
        metrics: AnalyzerMetrics,
        pulseState: EvolvePulseState
    ): SceneState {
        val now = System.currentTimeMillis()
        val accent = metrics.beatPulse * 0.55f + metrics.onsetPulse * 0.35f + metrics.attackMod * 0.1f
        pulseState.phraseMomentum = clamp(
            pulseState.phraseMomentum * 0.82f + metrics.energy * 0.12f + metrics.beatPulse * 0.22f + metrics.onsetPulse * 0.18f,
            0f,
            1.8f
        )

        val minInterval = 220f + (1f - accent) * 480f
        if (now - pulseState.lastMutationAt < minInterval.toLong()) {
            return scene
        }

        val mutationChance = accent * 0.9f + metrics.energy * 0.18f
        if (Random.nextFloat() > mutationChance) {
            return scene
        }

        pulseState.lastMutationAt = now

        val nextLayers = scene.enabledLayers.toMutableSet()
        val nextEffects = scene.enabledEffects.toMutableSet()
        var nextColorMode = scene.colorMode
        var nextFxIntensity = scene.fxIntensity
        var nextSpeed = scene.speed
        var nextTileCount = scene.tileCount
        var nextSymmetry = scene.symmetrySegments

        val phraseHit = pulseState.phraseMomentum > 1.15f && metrics.beatPulse > 0.58f
        val mutationCount = when {
            phraseHit -> randomInt(2, 4)
            accent > 0.72f -> randomInt(1, 3)
            else -> 1
        }

        repeat(mutationCount) {
            when (Random.nextFloat()) {
                in 0f..0.52f -> toggleLayer(nextLayers, 0.74f - metrics.decayMod * 0.2f)
                in 0.52f..0.8f -> toggleEffect(nextEffects, 0.58f + metrics.highs * 0.16f)
                in 0.8f..0.9f -> {
                    if (phraseHit || Random.nextFloat() < metrics.highs * 0.5f) {
                        nextColorMode = ColorMode.entries[randomInt(0, ColorMode.entries.size - 1)]
                    }
                }
                else -> {
                    nextFxIntensity = clamp(
                        nextFxIntensity + (metrics.attackMod - metrics.decayMod) * 0.08f + randomFloat(-0.04f, 0.04f),
                        0f,
                        1f
                    )
                    nextSpeed = clamp(
                        nextSpeed + metrics.beatPulse * 0.05f - metrics.decayMod * 0.03f + randomFloat(-0.03f, 0.03f),
                        0f,
                        1f
                    )
                    nextTileCount = (nextTileCount + if (phraseHit) randomInt(-1, 1) else 0).coerceIn(2, 6)
                    nextSymmetry = (nextSymmetry + if (metrics.highs > 0.62f && Random.nextFloat() < 0.4f) randomInt(-1, 1) else 0).coerceIn(4, 12)
                }
            }
        }

        if (phraseHit) {
            pulseState.phraseMomentum *= 0.58f
        }

        if (nextLayers.isEmpty()) {
            nextLayers.add(VisualLayer.entries[randomInt(0, VisualLayer.entries.size - 1)])
        }

        return scene.copy(
            enabledLayers = nextLayers,
            enabledEffects = nextEffects,
            colorMode = nextColorMode,
            fxIntensity = (nextFxIntensity * 100f).roundToInt() / 100f,
            speed = (nextSpeed * 100f).roundToInt() / 100f,
            tileCount = nextTileCount,
            symmetrySegments = nextSymmetry
        )
    }

    fun nextEvolveDelayMs(metrics: AnalyzerMetrics): Long {
        val accent = clamp(metrics.beatPulse * 0.6f + metrics.onsetPulse * 0.25f + metrics.energy * 0.15f, 0f, 1f)
        return (180 + (1f - accent) * 320f).roundToInt().toLong()
    }
}
